R/submit_ipr.R

```diff
@@ -1,145 +1,417 @@
+  ### global
+  tb_status <<- tibble::new_tibble(
+    list(
+      input_file = c(NULL),
+      job_id = c(NULL),
+      status = c(NULL),
+      polls = c(NULL),
+      t_submit = c(NULL),
+      t_latest_poll = c(NULL)
+    )
+  )
+  return(tb_status)
+}
+
+.add_job_status_table <- function(input_filename, job_code, submit_status) {
+  new_job_row <- tibble::as_tibble_row(
+    list(
+      input_file = input_filename,
+      job_id = job_code,
+      status = submit_status,
+      polls = 0L,
+      t_submit = as.POSIXct(Sys.time()),
+      t_latest_poll = as.POSIXct(Sys.time())
+    )
+  )
+  # coerce tb_status to prevent type errors on bind_rows (not sure why, but sometimes it is converted to a double)
+  tb_status[["t_latest_poll"]] <<- as.POSIXct(tb_status[["t_latest_poll"]])
+
+  tb_status <<- dplyr::bind_rows(tb_status, new_job_row)
+  return(NULL)
+  # return(job_code)
+}
 
-  ### POST
-  url_post <- 'https://www.ebi.ac.uk/Tools/services/rest/iprscan5/run'
+.submit <- function(path2seq, user_email) {
   fasta <- Biostrings::readAAStringSet(path2seq)
-  fasta_header <- names(fasta)[[1]]
   sequence_str <- as.character(fasta[[1]])
-  print(fasta)
-  headers_post <- c("Accept: text/plain", "Content-Type: application/x-www-form-urlencoded")
 
-  # job params
+  url_post <- "https://www.ebi.ac.uk/Tools/services/rest/iprscan5/run"
+  headers_post <- c("Accept: text/plain", "Content-Type: application/x-www-form-urlencoded")
   data <- list(
-    email=user_email,
-    title="test",
-    goterms="false",
-    pathways="false",
-    stype="p",
-    appl="PfamA",
-    sequence=sequence_str
+    email = user_email,
+    title = "riprscan",
+    goterms = "false",
+    pathways = "false",
+    stype = "p",
+    appl = c("PfamA", "CDD"),
+    sequence = sequence_str
   )
-
+  ### global
+  total_jobs <<- total_jobs + 1
   res_post <- httr::POST(url_post, body = data, httr::add_headers(headers_post))
-  cat("Request: \n", sep = "")
+  job_id <- rawToChar(res_post$content)
+  cat("Request info for ", job_id, ": \n", sep = "")
   print(res_post$request)
-  if (as.integer(res_post$status) != 200L) {
-    message <- paste0("POST error with sequence: ", fasta_header, "\n",
-                      "Status code: ", as.integer(res_post$status), "\n")
+
+  status <- ifelse(as.integer(res_post$status) == 200L, "running", "failed")
+  if (status == "failed") {
+    message <- paste0(
+      "POST error with sequence: ", fasta_header, "\n",
+      "Status code: ", as.integer(res_post$status), "\n"
+    )
     warning(message)
-    return(NULL)
-  }
-  else {
-    cat("POST status: ", as.integer(res_post$status), "\n", sep = "")
+    .add_job_status_table(basename(path2seq), job_id, status)
+    return("failed")
+  } else {
+    .add_job_status_table(basename(path2seq), job_id, status)
+    return(job_id)
   }
-  job_id <- rawToChar(res_post$content)
-  cat("job_id: ", job_id, "\n", sep = "")
+}
+
 
-  ### GET
+.poll_job <- function(job_code) {
```

Comment:
Jake, can you tell me a bit about the `poll` logic -- why/what are we polling here? Are we using this because iprscan api documentation uses this? Maybe we discussed this but I meant to ask yesterday & forgot!

Reply:
From my understanding, polling is just a term used for repeated requests at certain intervals in the context of web APIs. 

So, when I "poll" a job, im attempting to write the job output if it's available. If not, then I will "poll" again at some patterned interval.

Maybe, @falquaddoomi can clarify if this is the appropriate term to use here or not.

Reply:
That's correct, @jakekrol, although I'd add that polling is a general term for checking some condition at a regular interval, versus being notified when the condition changes. Also, as far as I can tell that's the correct way to get results from the Interproscan API; their Python implementation does the same thing.

Reply:
Thanks, both! Just wanted to make sure the vocabulary is right since we use `poll` in several functions here.